10161: watch for changing base during Mul flattening

smichr opened the pull request:
fixes #10161

smichr committed:
10161: watch for changing base during Mul flattening

asmeurer commented:
All the Travis builds are erroring with "No output has been received in the last 10 minutes, this potentially indicates a stalled build or something wrong with the build itself." I'm guessing this means there is a nontrivial performance hit here.

smichr commented:
> nontrivial performance hit

:-)  not really, just a garden-variety infinite loop that I have to watch out for

smichr commented on sympy/core/mul.py:
```diff
@@ -402,14 +402,30 @@ def _gather(c_powers):
+                    if p.is_Number:
+                        coeff *= p
+                        continue
+                b = bi = p
```
put this into `else` since `e == 1`

smichr commented on sympy/core/mul.py:
```diff
@@ -402,14 +402,30 @@ def _gather(c_powers):
+                p = Pow(p, e)
+                if p.is_Pow:
+                    b, e = p.as_base_exp()
+                    if b != bi:
```
add 'and not bi.is_Pow` to avoid unevaluated powers (and hopefully fix the single failing test)

smichr committed:
_parallel_dict_from_expr_no_gens aware of unevaluated Mul

Thanks to Kalevi Suominen (@jksuom) for pointing out the
need to handle repeated bases in this routine.

smichr commented:
I think this is ready to look at if the tests pass.

smichr commented:
@jksuom, I acknowleged you in the 2nd commit -- is that OK or would you like to be the author of that commit?

jksuom commented:
That is (more than) OK, thank you. Besides, you found a way of improving the code :-)
(Only, the test would rather belong to ``test__parallel_dict_from_expr_no_gens``.)

smichr commented:
Got the test correction, @jksuom . Does the rest looks non-invasive enough to core?

jksuom commented:
I hope there would be tests where the part using ``as_base_exp`` were exercised. I find it hard to see in which cases the loop would have to be repeated and what would happen then to ``coeff``.

smichr commented:
> I hope there would be tests where the part using as_base_exp were exercised

That is what the 10161 test is using. The loop needs to be re-run to gather the `x` and the new `x**2` (from `abs(x)**2`) from `x*abs(x)*abs(x)`. As for the coefficient being updated...that was code that was presumably needed if the loop runs once but I didn't check that for coverage; it probably won't be hit while running it the second time. I'll try to see how it's covered later today.

jksuom commented:
I have taken some more time studying the code. It looks a bit intricate on the surface but that probably cannot be helped. Also, I now see that ``coeff`` will be updated only once as it is not appended to ``new_c_powers``.
On the whole, the modifications are quite conservative. Only one repetition is allowed. (That can be changed later if necessary, but I think that is improbable.) I don't think there would be noticeable effect on performance. I am +1 for this.

smichr commented:
Thanks, @jksuom , for looking this over. I'll go ahead and commit it as is.
